src/main/cpp/src/HostTableJni.cpp

```diff
@@ -0,0 +1,249 @@
+  });
+  return std::make_unique<host_table_view>(cols);
+}
+
+cudf::column_view to_device_column(host_column_view const& host_col, jlong host_to_dev_offset)
+{
+  auto data = host_col.data<uint8_t>();
+  if (data != nullptr) { data += host_to_dev_offset; }
+  auto mask = host_col.null_mask();
+  if (mask != nullptr) { mask += host_to_dev_offset / sizeof(*mask); }
+  std::vector<cudf::column_view> children;
+  std::transform(host_col.child_begin(),
+                 host_col.child_end(),
+                 std::back_inserter(children),
+                 [host_to_dev_offset](host_column_view const& c) {
+                   return to_device_column(c, host_to_dev_offset);
+                 });
+  return cudf::column_view(
+    host_col.type(), host_col.size(), data, mask, host_col.null_count(), 0, children);
+}
+
+std::vector<std::unique_ptr<cudf::column_view>> to_device_column_views(
+  host_table_view const& host_table, jlong host_to_dev_offset)
+{
+  std::vector<std::unique_ptr<cudf::column_view>> cv_ptrs;
+  cv_ptrs.reserve(host_table.num_columns());
+  std::transform(
+    host_table.begin(),
+    host_table.end(),
+    std::back_inserter(cv_ptrs),
+    [host_to_dev_offset](host_column_view const& host_col) {
+      return std::make_unique<cudf::column_view>(to_device_column(host_col, host_to_dev_offset));
+    });
+  return cv_ptrs;
+}
+
+}  // anonymous namespace
+
+extern "C" {
+
+JNIEXPORT jlong JNICALL Java_com_nvidia_spark_rapids_jni_HostTable_bufferSize(JNIEnv* env,
+                                                                              jclass,
+                                                                              jlong table_handle,
+                                                                              jlong jstream)
+{
+  JNI_NULL_CHECK(env, table_handle, "table is null", 0);
+  try {
+    cudf::jni::auto_set_device(env);
+    auto t      = reinterpret_cast<cudf::table_view const*>(table_handle);
+    auto stream = reinterpret_cast<cudaStream_t>(jstream);
+    return static_cast<jlong>(host_buffer_size(*t, stream));
+  }
+  CATCH_STD(env, 0);
+}
+
+JNIEXPORT jlong JNICALL Java_com_nvidia_spark_rapids_jni_HostTable_copyFromTableAsync(
+  JNIEnv* env, jclass, jlong table_handle, jlong host_address, jlong host_size, jlong jstream)
+{
+  JNI_NULL_CHECK(env, table_handle, "table is null", 0);
+  try {
+    cudf::jni::auto_set_device(env);
+    auto table           = reinterpret_cast<cudf::table_view const*>(table_handle);
+    auto buffer          = reinterpret_cast<uint8_t*>(host_address);
+    auto buffer_size     = static_cast<std::size_t>(host_size);
+    auto stream          = reinterpret_cast<cudaStream_t>(jstream);
+    auto host_table_view = to_host_table_async(*table, buffer, buffer_size, stream);
+    return reinterpret_cast<jlong>(host_table_view.release());
+  }
+  CATCH_STD(env, 0);
+}
+
+JNIEXPORT jlongArray JNICALL Java_com_nvidia_spark_rapids_jni_HostTable_toDeviceColumnViews(
```

Comment:
Does this mean we are creating `cudf::column_view` from host data pointer? I see that later we call `ColumnVector.fromViewWithContiguousAllocation(columnViewHandles)` on the output of this, but does that  implicitly handle host-device memcpy?

Reply:
We're creating a cudf::column_view from a host view but also applying a constant offset to each pointer that effectively translates the pointers from host memory into device memory,  See https://github.com/NVIDIA/spark-rapids-jni/pull/2393/files#diff-19b375574bb1514d837e2a4fbf847c446512d330b685cf12fa67f33ea9978d1fR116-R118 for details.  The host buffer is copied to device memory and the difference in base pointers of the host and device buffers is used to translate the host views into device views.